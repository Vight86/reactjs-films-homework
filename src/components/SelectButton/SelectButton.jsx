import React from 'react';
import propTypes from 'prop-types';
import Button from '../Button/index';
import styles from './SelectButton.scss';

const SelectButton = ({ genres }) => {
  const genreList = genres.map((genre) => (
    <li key={genre} className={styles.selectOption} data-value={genre}>{genre}</li>
  ));

  return (

    <span className={styles.selectButton}>
      <Button className="filterButton">Genre</Button>
      <span className={styles.selectArrow} />
      <ul className={styles.selectOptionsList}>
        <li className={styles.selectOption} data-value="Genre">Genre</li>
        {genreList}
      </ul>
    </span>
  );
};

SelectButton.propTypes = {
  genres: propTypes.arrayOf(propTypes.string),
};

SelectButton.defaultProps = {
  genres: ['', ''],
};

export default SelectButton;
