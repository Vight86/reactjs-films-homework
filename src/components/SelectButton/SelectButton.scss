@import '../../utils/variables';

.selectOptionsList {
  position: absolute;
  z-index: 1000;
  left: 0;
  width: max-content;
  padding: 0;
  margin: 0;
  list-style: none;
  opacity: 0;
  pointer-events: none;
  transition: all 0.3s;
  visibility: hidden;
}

.selectArrow {
  position: absolute;
  top: 1em;
  right: 3em;
  width: 13px;
  height: 13px;
  transform: rotate(180deg);
  transform-origin: center center;
  transition: transform 0.3s;

  &::before {
    position: absolute;
    top: 4px;
    left: -2px;
    display: inline-block;
    width: 10px;
    height: 3px;
    background-color: $color-secondary;
    border-radius: 6px;
    content: "";
    transform: rotate(-38deg);
  }

  &::after {
    position: absolute;
    top: 4px;
    right: -2px;
    display: inline-block;
    width: 10px;
    height: 3px;
    background-color: $color-secondary;
    border-radius: 3px;
    content: "";
    transform: rotate(38deg);
  }
}

.selectButton {
  position: relative;
  display: inline-block;

  &.active {
    & > .selectOptionsList {
      opacity: 1;
      pointer-events: all;
      visibility: visible;
    }

    & > .selectArrow {
      transform: rotate(0);
    }
  }
}

.selectOption {
  padding: 10px;
  border-bottom: 1px solid transparentize($color: $text-tertiary, $amount: 0.9);
  background-color: $color-light;

  & > button {
    width: 100%;
    height: 100%;
    border: none;
    background: none;
    color: $color-secondary;
    cursor: pointer;
    font-family: $font-primary;
    font-size: 2.8em;
    outline: none;
    text-align: left;

    &:hover {
      font-weight: 700;
    }
  }
}
