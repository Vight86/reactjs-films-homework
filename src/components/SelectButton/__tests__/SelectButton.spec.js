import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import SelectButton from '../index';

const mockGenres = [
  { id: 1, name: 'Action' },
  { id: 2, name: 'Scifi' },
];
const mockFiterMovies = jest.fn(() => 'tested');

let renderer;
let result;
beforeEach(() => {
  renderer = new ShallowRenderer();
  renderer.render(
    <SelectButton
      genres={mockGenres}
      filterMovies={mockFiterMovies}
    >
      Title
    </SelectButton>,
  );
  result = renderer.getRenderOutput();
});

afterEach(() => {
  renderer = null;
  result = null;
});

describe('render SelectButton component', () => {
  it('render component correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('render handleClick correctly', () => {
    const selectOptionButton = result.props.children[2].props.children[1][1].props.children;
    const mockEvent = {
      target: {
        innerHTML: 'tested',
        dataset: {
          value: '2',
        },
      },
    };
    selectOptionButton.props.onClick(mockEvent);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render filterButton opening click correctly', () => {
    const filterButton = result.props.children[0];
    filterButton.props.onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
