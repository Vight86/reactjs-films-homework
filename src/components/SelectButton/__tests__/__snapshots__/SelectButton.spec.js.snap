// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SelectButton component render component correctly 1`] = `
<span
  className="selectButton"
>
  <Button
    className="filterButton"
    onClick={[Function]}
  >
    Title
  </Button>
  <span
    className="selectArrow"
  />
  <ul
    className="selectOptionsList"
  >
    <li
      className="selectOption"
    >
      <button
        data-value=""
        onClick={[Function]}
        type="button"
      >
        All genres
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={1}
        onClick={[Function]}
        type="button"
      >
        Action
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={2}
        onClick={[Function]}
        type="button"
      >
        Scifi
      </button>
    </li>
  </ul>
</span>
`;

exports[`render SelectButton component render filterButton opening click correctly 1`] = `
<span
  className="selectButton active"
>
  <Button
    className="filterButton"
    onClick={[Function]}
  >
    Title
  </Button>
  <span
    className="selectArrow"
  />
  <ul
    className="selectOptionsList"
  >
    <li
      className="selectOption"
    >
      <button
        data-value=""
        onClick={[Function]}
        type="button"
      >
        All genres
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={1}
        onClick={[Function]}
        type="button"
      >
        Action
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={2}
        onClick={[Function]}
        type="button"
      >
        Scifi
      </button>
    </li>
  </ul>
</span>
`;

exports[`render SelectButton component render handleClick correctly 1`] = `
<span
  className="selectButton active"
>
  <Button
    className="filterButton"
    onClick={[Function]}
  >
    tested
  </Button>
  <span
    className="selectArrow"
  />
  <ul
    className="selectOptionsList"
  >
    <li
      className="selectOption"
    >
      <button
        data-value=""
        onClick={[Function]}
        type="button"
      >
        All genres
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={1}
        onClick={[Function]}
        type="button"
      >
        Action
      </button>
    </li>
    <li
      className="selectOption"
    >
      <button
        data-value={2}
        onClick={[Function]}
        type="button"
      >
        Scifi
      </button>
    </li>
  </ul>
</span>
`;
